feat(appsync): import existing graphql api

BryanPan342 opened the pull request:
**[ISSUE]**
Appsync missing `fromXxx` functionality, making multiple stack appsync interaction impossible. 

**[Approach]**
Created a base class for  `GraphQLApi` and a `IGraphQLApi` interface for imports. Added `fromXxxAttributes` functions to code base that can add dataSources.

**[Notes]**
Only accessible props from `IGraphQLApi` are `apiId` and `arn`. Only accessible functions from `IGraphQLApi` are the `addXxxDataSource` functions.

Added `props-physical-name:@aws-cdk/aws-appsync.GraphQLApiProps` as linter exception because the breaking change isn't worth the return of making the physical name (name exists already).

Added `from-method:@aws-cdk/aws-appsync.GraphQLApi` as linter exception because a `fromGraphQLApiAttributes` function will turn into `from_graph_q_l_api_attributes` in python.

Fixes #6959 

BREAKING CHANGE: **appsync.addXxxDataSource** `name` and `description` props are now optional and in an `DataSourceOptions` interface. 
- **appsync**:  the props `name` and `description` in `addXxxDataSource` have been moved into new props `options`  of type `DataSourceOptions`
- **appsync**: `DataSourceOptions.name` defaults to id
- **appsync**: `DataSourceOptions.description` defaults to undefined

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3bde2e92e0e225f27c553bf37ca202631ac6bdcc
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fdbf05f3-8691-4b5a-90c0-132232edf644%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 48f782481b1ef41db9f972c79d48ff01b19752a8
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=27461185-f860-4d58-a647-afae42831268%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a68e1f655d63de79530055dfd9aa17b268d1f6c4
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e62c2ecc-4560-4df2-bab3-d1ed16c560d9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:
```diff
@@ -300,11 +297,71 @@ export class IamResource {
+    class Import extends GraphQLApiBase {
+      public readonly apiId = attrs.apiId;
+      public readonly arn = attrs.arn;
+      constructor (s: Construct, i: string){
+        super(s, i);
```
avoid abbreviating - it's more readable to spell out scope and id

BryanPan342 replied:
i cant do scope and id again because it is already declared in as params for the function itself

shivlaks replied:
gotcha, let's leave it for now in that case

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:
```diff
@@ -315,6 +372,8 @@ export class GraphQLApi extends Construct {
+   *
+   * @attribute
    */
   public readonly graphQlUrl: string;
```
curious about the casing here (and everywhere really)... how does this translate to Python where we do snake casing?

does `GraphQL` get broken down into `graph_q_l`?

BryanPan342 replied:
yeah i think in this case would be best to change all of them to graphqlUrl for props

BryanPan342 replied:
^ this would be a breaking change tho so not sure if thats the best move for something so small..

@MrArnoldPalmer  thoughts?

shivlaks replied:
I'm not sure it needs to happen in this PR, but fixing the casing sounds like something we need to do regardless if it requires code that is not idiomatic for Python developers

BryanPan342 replied:
yeah i think a chore might be better

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:
```diff
@@ -597,6 +590,15 @@ export class GraphQLApi extends Construct {
+   * Add schema dependency if not imported
+   * @param construct the construct that has a dependency
+   */
+  public addSchemaDependency( construct: CfnResource ): boolean {
```
```suggestion
  public addSchemaDependency(construct: CfnResource ): boolean {
```

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:
```diff
@@ -597,6 +590,15 @@ export class GraphQLApi extends Construct {
+   * @param construct the construct that has a dependency
+   */
+  public addSchemaDependency( construct: CfnResource ): boolean {
+    construct.addDependsOn(this.schema);
+    return true;
```
do we need this method? it seems like it's substituting one API call (`addDependsOn`) with another (`addSchemaDependency`). any reason users can't just do it directly?

does it require validation? what's valid as a construct to add

BryanPan342 replied:
there needs to be a dependency add for resolvers and the schema, but i was not comfortable passing schema's around in imports so essentially i needed a way to allow for adding a dependency only if the api wasn't imported. 

this basically was my workaround.. if anything i can make this function protected

shivlaks replied:
can  we make it internal?

BryanPan342 replied:
actually, I've thought about this a little more and I think that it might be good to offer a public api to add a dependency.. if the graphql api is in another stack, it will be deployed before the imported stack so the dependency doesn't matter too much? 

but in the case that it does we should have a public api to allow for adding dependencies? if not i dont see why a public api is a net bad

shivlaks replied:
if it enables use cases that users cannot achieve, then I agree.
this method calls `construct.addDependsOn(schema)` and returns true. can users just make that api call instead?

BryanPan342 replied:
@shivlaks its just really annoying for users to have to do after creating each resolver. 

The only reason its public as of right now is that I need `IGraphQLApi` to have some capability to reference a schema, I can't make it protected or the function will be inaccessible to the Resolver class.

shivlaks commented on packages/@aws-cdk/aws-appsync/package.json:
```diff
@@ -94,7 +94,8 @@
       "no-unused-type:@aws-cdk/aws-appsync.ApiKeyConfig",
       "no-unused-type:@aws-cdk/aws-appsync.UserPoolConfig",
-      "no-unused-type:@aws-cdk/aws-appsync.UserPoolDefaultAction"
+      "no-unused-type:@aws-cdk/aws-appsync.UserPoolDefaultAction",
+      "props-physical-name:@aws-cdk/aws-appsync.GraphQLApiProps"
```
include rationale for any/all added linter exclusions in the commit body

BryanPan342 commented on packages/@aws-cdk/aws-appsync/README.md:
````diff
@@ -81,6 +81,23 @@ demoDS.createResolver({
+importedApi.addDynamoDbDataSource(...);
+```
+
+GraphQL Apis can be imported in three ways:
+- `arn` through `GraphQLApi.fromGraphQLApiArn`: `apiId` is generated from parameter
+- `apiId` through `GraphQLApi.fromGraphQLApiId`: `arn` is generated from parameter
+- or both through `GraphQLApi.fromGraphQLApiAttributes`: both are specified
````
in this case then i think it might just better to do `attributes` and `ApiId` because as you said it doest get confusing.

BryanPan342 replied:
actually i take that back.. it would be better to just to `attributes` and make the arn optional..

the arn isn't that necessary to be honest in terms of creating resolvers/data sources.. the only thing it would be useful would be for grants

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,142 @@
+   * add a new DynamoDB data source to this API
+   * @param name The name of the data source
+   * @param description The description of the data source
+   * @param table The DynamoDB table backing this data source [disable-awslint:ref-via-interface]
```
o this was copy pasta from before i think we can remove this now because of the a PR that allowed for ITable input

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,142 @@
+   * @attribute
+   */
+  readonly arn: string;
+  /**
+   * add a new dummy data source to this API
+   * @param name The name of the data source
+   * @param description The description of the data source
+   */
+  addNoneDataSource(name: string, description: string): NoneDataSource;
```
> in this case it's when you want to create a mutation that's triggered by the backend, etc. it does not require a data source.

yup this is the case. ill add it to the doc string

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,142 @@
+   * @param description The description of the data source
+   * @param table The DynamoDB table backing this data source [disable-awslint:ref-via-interface]
+   */
+  addDynamoDbDataSource( name: string, description: string, table: ITable ): DynamoDbDataSource;
```
Why is that? there aren't that many configurable parameters so i'm not sure if an `options` prop is necessary

shivlaks replied:
it's consistent with the rest of the construct library and also compliant with [design guidelines](https://github.com/aws/aws-cdk/blob/master/DESIGN_GUIDELINES.md)

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,142 @@
+
+  /**
+   * Add schema dependency if not imported
+   * @param construct the construct that has a dependency
```
construct is the dependee

shivlaks replied:
can we provide a more helpful description.
what happens if this method is called and it is imported? it's unclear what the implications are if it is imported

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,142 @@
+ */
+export interface IGraphQLApi extends IResource {
+  /**
+   * the id of the GraphQL API
```
its a unique id so i put an example of one that i get but its honestly kinda gibberish

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b0ea1a23a1fc2b301e4901da6b333cf179a78aaa
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c4642041-559b-4427-9b23-4935576b78f2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d1e7c48296237402a513324be6072f9d1319eb50
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a3a5b223-70f5-4ec5-aea9-6f7836088ca7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,146 @@
+
+  /**
+   * add a new none data source to this API. Useful for pipeline resolvers
+   * and for backend changes that don't require a data source.
+   * @param name The name of the data source @default 'None'
```
nit: add  a blank line between the API description and the `@param` here (and elsewhere in the PR)

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,146 @@
+
+  /**
+   * an unique AWS AppSync GraphQL API identifier
+   * i.e. 'lxz775lwdrgcndgz3nurvac7oa'
```
nice! can you add proliferate this everywhere in this PR where you have a doc string for API id?

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,146 @@
+   * add a new none data source to this API. Useful for pipeline resolvers
+   * and for backend changes that don't require a data source.
+   * @param name The name of the data source @default 'None'
+   * @param description The description of the data source @default undefined
```
move the `@default` to it's own line

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:
```diff
@@ -300,31 +297,76 @@ export class IamResource {
+   * @param id id
+   * @param attrs GraphQL API Attributes of an API
+   */
+  public static fromGraphQLApiAttributes(scope: Construct, id: string, attrs: GraphQLApiAttributes): IGraphQLApi {
```
same comment re: casing - although we don't need to clean every scenario up, we should not proliferate something we know won't be idiomatic in Python. let's go for what we want to end up with at the outset instead of creating future chore work.

shivlaks commented on packages/@aws-cdk/aws-appsync/test/appsync-import.test.ts:
```diff
@@ -0,0 +1,133 @@
+let stack: cdk.Stack;
+let baseApi: appsync.GraphQLApi;
+beforeEach( () => {
+  // Given
```
```suggestion
  // GIVEN
```

shivlaks commented on packages/@aws-cdk/aws-appsync/test/integ.api-import.ts:
```diff
@@ -0,0 +1,67 @@
+ * Install dependencies and deploy integration test. Check if data sources are
+ * connected to the graphQL Api
+ *
+ * -- cdk deploy --app 'node integ.api-import.js' stack            -- start         --
+ * -- aws appsync list-graphql-apis                                -- obtain api id --
+ * -- aws appsync list-data-sources --api-id [api_id]              -- testDS/None   --
+ * -- cdk destroy --app 'node integ.api-import.js' stack baseStack -- clean         --
```
nit: i don't think the deploy/destroy need to be covered - just the steps to verify the expected configuration.

BryanPan342 replied:
there are two stacks so maybe its important? im fine with removing it though

shivlaks replied:
it's a nit, so up to you!

shivlaks commented on packages/@aws-cdk/aws-appsync/test/integ.api-import.ts:
```diff
@@ -0,0 +1,67 @@
+/// !cdk-integ *
```
is this needed?

BryanPan342 replied:
yes because there are two stacks

shivlaks replied:
got it

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,146 @@
+   * @param description The description of the data source
+   * @param table The DynamoDB table backing this data source
+   */
+  addDynamoDbDataSource(name: string, description: string, table: ITable): DynamoDbDataSource;
```
it sounds like name and description should be optional for any given data source? we should be able to create sensible defaults if they're required under the hood.

I'm of the opinion that the signature should require a table and options to set the rest... but I'll defer to whatever @MrArnoldPalmer thinks

BryanPan342 replied:
we talked offline and cleared up the confusion. I thought the implementation would end up looking like this:

```ts
api.addDynamoDbDataSource('tableDs', {
  table: table,
  description: 'my table',
});
```

whereas @shivlaks was suggesting this:

```ts
api.addDynamoDbDataSource(table, {
  name: 'tableDs',
  description: 'my table',
});
```

where the `options` is optional. I agree, this is MUCH cleaner

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 142c5f5145b81d51ee5f96247bb8ce93d11ea21d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=52ff9d95-4a38-431f-b52b-547720f8d710%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7e9b775995511bfdbb1c27309d754bc96c3de67e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1c560331-1933-46c6-9f4b-840331237b84%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a44a9b883811098fe32bf237534fb76c9a7677ef
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c0b4890d-db53-41b2-b387-2a6804cfdb21%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

BryanPan342 commented:
Wow this PR got fat real quick :( 

To make reviewing easier, I will summarize and highlight important files:

FILES ADDRESSED BY BREAKING CHANGES (not that important):
- **integ.graphl-iam.expected.json**
- **integ.graphl-iam.ts**
- **integ.graphl.expected.json**
- **integ.graphl.ts**

INTEG TESTS:
- **integ.api-import.expected.json**
- **integ.api-import.ts**

**[NEW]** UNIT TEST FILES (mostly the same test but across different data sources):
- **appsync-dynamodb.test.ts**
- **appsync-none.test.ts**
- **appsync-http.test.ts**
- **appsync-lambda.test.ts**

LOGIC (most important):
- **[NEW]** **graphqlapi-base.ts**: added an interface `DataSourceOptions` and changed implementation in base class to create defaults
- **graphqlapi.ts**: unchanged
- **datasource.ts**: unchanged
- **resolver.ts** : unchanged

---

We are still discussing the topic of adding a dependency for resolvers to schema. https://github.com/aws/aws-cdk/blob/a44a9b883811098fe32bf237534fb76c9a7677ef/packages/%40aws-cdk/aws-appsync/lib/graphqlapi-base.ts#L90

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * The description of the data source
+   *
+   * @default - undefined
```
```suggestion
   * @default - No description
```

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * The name of the data source
+   *
+   * @default - '<DataSourceType>CDKDefault'
```
perhaps rephrase to`Automatically generated name` and include an example.
I'm not super fond of the default part but I agree CDK should be in the default.

can multiple data sources be added with the same "friendly name" or will that complain? Can I just add 2 data sources of the same type with the defaults?

BryanPan342 replied:
No, i created a unit test to make sure of this. But essentially each resolver is a construct and a construct must have a unique name.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d7678bab9310ad14c5fb37dd1b0484b920cb4da7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e9585b2f-44f1-4e40-b94d-e839df19f79b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

BryanPan342 commented on packages/@aws-cdk/aws-appsync/README.md:
```diff
@@ -82,6 +82,24 @@ demoDS.createResolver({
+  graphqlApiId: api.apiId,
+  graphqlArn: api.arn,
+});
+importedApi.addDynamoDbDataSource('TableDataSource', 'Table', table);
```
```suggestion
importedApi.addDynamoDbDataSource(table, 'tableId');
```
and change everywhere else

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c066ba9c60fdae5285ec86e6e3e5503f375e1d95
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=66ae4a32-082e-4c61-92a3-07526054d29f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ecc91452c31cc55114438f6936baf2b622faa434
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5cb53497-7248-4cdb-9123-5ce5c2c63d34%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 4976bab72cbd8e531806395d1a2d5f1be52287c8
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=365d73e9-422d-43fd-bee6-ff0aa1e66000%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5243ef0732f41f61c813ac9c02d82d0f0e682ed0
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=221e31e5-74e3-48c4-b0a1-3f19d28001b6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,193 @@
+  /**
+   * The name of the data source, overrides the id given by cdk
+   *
+   * @default - the id provisioned
```
```suggestion
   * @default - the id provisioned
```

let's make this a little more helpful and say `Generated by ` ...

BryanPan342 replied:
wait but the name is just the id for now..

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,193 @@
+   *
+   * @param id The data source's id
+   * @param options The optional configuration for this data source
+   * @default name - the id
```
I don't believe `@default` is needed for these methods - but if you did include it, describe the default behavior from not providing any `options`

BryanPan342 replied:
sweet okay i can remove them

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,193 @@
+   * @param table The DynamoDB table backing this data source
+   * @param options The optional configuration for this data source
+   * @default name - the id
+   * description - No description
```
I believe this should be indented by a couple spaces so that it's clear that this comment belongs to the `@default` 
we don't have too many examples that span beyond a single line, but that's the guidance I was given in the past :)

BryanPan342 replied:
im just removing them

shivlaks commented on packages/@aws-cdk/aws-appsync/lib/graphqlapi-base.ts:
```diff
@@ -0,0 +1,193 @@
+  /**
+   * the ARN of the API
+   */
+  public abstract readonly arn: string;
```
did we want to give this a specific prefix before the `arn` - since we're introducing this, let's make sure it aligns with how we define arn elsewhere

BryanPan342 replied:
if we change this hear i feel like we need to change it for every where else.. including the GraphQLAPI.. i think this breaking change should happen in another pr

shivlaks commented on packages/@aws-cdk/aws-appsync/test/appsync-dynamodb.test.ts:
```diff
@@ -1,15 +1,93 @@
+    // WHEN
+    api.addDynamoDbDataSource('ds', table);
+
+    // EXPECT
```
```suggestion
    // THEN
```

rationale: align with our convention elsewhere. few places in this PR that need to align to the convention

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a668e1304de6fb6eaeff27f8c15faddb7f616789
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4d2047ae-a493-47bf-9dac-4c614d099a1d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5a15cf067bc4e89766472db0f1ae6843e1ed5e33
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9176fc2a-2f08-4bf8-aebf-49b7c3d3afef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
